better boxing

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30370 [wip] Use codegen'ed unboxing wrappers
* #29934 Boxed variable dispatch
* #29682 backend fallback test
* **#29681 better boxing**
* #29337 Hide the OperatorKernel* argument from the stack based kernel API
* #29330 OperatorHandle::callBoxed/callUnboxed
* #29214 Test cases for backend fallback kernels
* #29201 Add a OperatorHandle argument to boxed kernels
* #30340 Make Dispatcher::backendFallbackKernels_ an array
* #30333 [pytorch][PR] Remove LeftRight from OperatorEntry and DispatchTable.
* #30332 [pytorch][PR] Convert KernelTable to a flat-indexed array rather than a hashtable.

Remove callUnboxedOnly() and instead use metaprogramming to figure out if an operator can use a boxed fallback or not.
This enables boxed fallback for ops in native_functions.yaml even if they don't have `use_c10_dispatcher: full` set, as long as they're in the range of supported types.

Differential Revision: [D18462653](https://our.internmc.facebook.com/intern/diff/D18462653/)

smessmer committed:
better boxing

Remove callUnboxedOnly() and instead use metaprogramming to figure out if an operator can use a boxed fallback or not.
This enables boxed fallback for ops in native_functions.yaml even if they don't have `use_c10_dispatcher: full` set, as long as they're in the range of supported types.

Differential Revision: [D18462653](https://our.internmc.facebook.com/intern/diff/D18462653/)

[ghstack-poisoned]

smessmer committed:
Update on "better boxing"


Remove callUnboxedOnly() and instead use metaprogramming to figure out if an operator can use a boxed fallback or not.
This enables boxed fallback for ops in native_functions.yaml even if they don't have `use_c10_dispatcher: full` set, as long as they're in the range of supported types.

Differential Revision: [D18462653](https://our.internmc.facebook.com/intern/diff/D18462653/)

[ghstack-poisoned]

ezyang commented:
There is a lot of irrelevant code motion in this PR, which makes it hard to read the diff. Could you please explain what happened?

ezyang commented on aten/src/ATen/core/boxing/boxing.h:
```diff
@@ -0,0 +1,80 @@
+  TORCH_INTERNAL_ASSERT(false, "Tried to call KernelFunction::callUnboxed() for a kernel that only has a boxed kernel and doesn't support calling from an unboxed API yet.");
+}
+
+template<class Result, class... Args>
```
A little comment saying which sfinae is for which case would help code readers

ezyang commented on aten/src/ATen/core/boxing/KernelFunction_impl.h:
```diff
@@ -0,0 +1,177 @@
+
+namespace c10 {
+
+inline KernelFunction::KernelFunction()
```
Are there any changes to this file?

smessmer replied:
`callUnboxed` now calls into `impl::boxAndCallBoxedFunc` instead of `detail::boxAndCallBoxedFunc` and that function implementation moved into its own header `boxing.h`. That's also the reason why this whole code movement happens - there's a dependency cycle between `KernelFunction.h` and `boxing.h` that I'm resolving by creating a `KernelFunction_impl.h`.

ezyang replied:
OK... but did you change any of the code in this file? :)

smessmer replied:
This file is new and it has all the `KernelFunction` methods. The code inside the methods is unchanged except for what I mentioned above.

smessmer committed:
Update on "better boxing"


Remove callUnboxedOnly() and instead use metaprogramming to figure out if an operator can use a boxed fallback or not.
This enables boxed fallback for ops in native_functions.yaml even if they don't have `use_c10_dispatcher: full` set, as long as they're in the range of supported types.

Differential Revision: [D18462653](https://our.internmc.facebook.com/intern/diff/D18462653/)

[ghstack-poisoned]

smessmer committed:
Update on "better boxing"


Remove callUnboxedOnly() and instead use metaprogramming to figure out if an operator can use a boxed fallback or not.
This enables boxed fallback for ops in native_functions.yaml even if they don't have `use_c10_dispatcher: full` set, as long as they're in the range of supported types.

Differential Revision: [D18462653](https://our.internmc.facebook.com/intern/diff/D18462653/)

[ghstack-poisoned]

smessmer committed:
Update on "better boxing"


Remove callUnboxedOnly() and instead use metaprogramming to figure out if an operator can use a boxed fallback or not.
This enables boxed fallback for ops in native_functions.yaml even if they don't have `use_c10_dispatcher: full` set, as long as they're in the range of supported types.

Differential Revision: [D18462653](https://our.internmc.facebook.com/intern/diff/D18462653/)

[ghstack-poisoned]

smessmer committed:
Update on "better boxing"


Remove callUnboxedOnly() and instead use metaprogramming to figure out if an operator can use a boxed fallback or not.
This enables boxed fallback for ops in native_functions.yaml even if they don't have `use_c10_dispatcher: full` set, as long as they're in the range of supported types.

Differential Revision: [D18462653](https://our.internmc.facebook.com/intern/diff/D18462653/)

[ghstack-poisoned]

smessmer committed:
Update on "better boxing"


Remove callUnboxedOnly() and instead use metaprogramming to figure out if an operator can use a boxed fallback or not.
This enables boxed fallback for ops in native_functions.yaml even if they don't have `use_c10_dispatcher: full` set, as long as they're in the range of supported types.

Differential Revision: [D18462653](https://our.internmc.facebook.com/intern/diff/D18462653/)

[ghstack-poisoned]

smessmer committed:
Update on "better boxing"


Remove callUnboxedOnly() and instead use metaprogramming to figure out if an operator can use a boxed fallback or not.
This enables boxed fallback for ops in native_functions.yaml even if they don't have `use_c10_dispatcher: full` set, as long as they're in the range of supported types.

Differential Revision: [D18462653](https://our.internmc.facebook.com/intern/diff/D18462653/)

[ghstack-poisoned]

smessmer committed:
Update on "better boxing"


Remove callUnboxedOnly() and instead use metaprogramming to figure out if an operator can use a boxed fallback or not.
This enables boxed fallback for ops in native_functions.yaml even if they don't have `use_c10_dispatcher: full` set, as long as they're in the range of supported types.

Differential Revision: [D18462653](https://our.internmc.facebook.com/intern/diff/D18462653/)

[ghstack-poisoned]

smessmer committed:
Update on "better boxing"


Remove callUnboxedOnly() and instead use metaprogramming to figure out if an operator can use a boxed fallback or not.
This enables boxed fallback for ops in native_functions.yaml even if they don't have `use_c10_dispatcher: full` set, as long as they're in the range of supported types.

Differential Revision: [D18462653](https://our.internmc.facebook.com/intern/diff/D18462653/)

[ghstack-poisoned]

smessmer committed:
Update on "better boxing"


Remove callUnboxedOnly() and instead use metaprogramming to figure out if an operator can use a boxed fallback or not.
This enables boxed fallback for ops in native_functions.yaml even if they don't have `use_c10_dispatcher: full` set, as long as they're in the range of supported types.

Differential Revision: [D18462653](https://our.internmc.facebook.com/intern/diff/D18462653/)

[ghstack-poisoned]

smessmer committed:
Update on "better boxing"


Remove callUnboxedOnly() and instead use metaprogramming to figure out if an operator can use a boxed fallback or not.
This enables boxed fallback for ops in native_functions.yaml even if they don't have `use_c10_dispatcher: full` set, as long as they're in the range of supported types.

Differential Revision: [D18462653](https://our.internmc.facebook.com/intern/diff/D18462653/)

[ghstack-poisoned]

smessmer committed:
Update on "better boxing"


Remove callUnboxedOnly() and instead use metaprogramming to figure out if an operator can use a boxed fallback or not.
This enables boxed fallback for ops in native_functions.yaml even if they don't have `use_c10_dispatcher: full` set, as long as they're in the range of supported types.

Differential Revision: [D18462653](https://our.internmc.facebook.com/intern/diff/D18462653/)

[ghstack-poisoned]
